### PR TITLE
support lambda layer

### LINKED ISSUE #703: Add support for deploying/managing lambda layers
**Is your feature request related to a problem? Please describe.**

I was running into a problem where I wanted to share code between two different lambda functions.

**Describe the solution you'd like**

I would like to easily manage/deploy/use/test with lambda layers in my AWS Amplify project to share code between two different lambda functions.
https://aws.amazon.com/blogs/aws/new-for-aws-lambda-use-any-programming-language-and-share-common-components/

**Describe alternatives you've considered**

I’ve thought about symlinking some shared files, but this feels really jenky and brittle.

**Additional context**

By making this part of the AWS Amplify CLI, I can easily have all my code managed in one clean way, and be sure that when I deploy things, everything has the latest up-to-date code running and is deployed in a clean manner.

### PR DESCRIPTION
**Pull request by hirochachacha**
Fixes #703

Follow-up PR of https://github.com/aws-amplify/amplify-cli/pull/3135

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-function/provider-utils/awscloudformation/index.js [hunk 1]
```diff
@@ -315,27 +315,32 @@ async function openEditor(context, category, options) {
 async function invoke(context, category, service, resourceName) {
   const { amplify } = context;
   serviceMetadata = amplify.readJsonFile(`${__dirname}/../supported-services.json`)[service];
-  const { inputs } = serviceMetadata;
+  const { inputs: rawInputs } = serviceMetadata;
+  const inputs = {};
+  rawInputs.forEach(input => {
+    inputs[input.key] = input;
+  });
+  const { handlerFilename, handlerName, eventName } = input;
```
Comment (rayjanwilson) [inline packages/amplify-category-function/provider-utils/awscloudformation/index.js]: should this be inputs, plural?
`const { handlerFilename, handlerName, eventName } = inputs;`

like it is in the file `packages/amplify-category-function/provider-utils/awscloudformation/service-walkthroughs/lambda-walkthrough.js` line 25?

Comment (hirochachacha) [reply packages/amplify-category-function/provider-utils/awscloudformation/index.js]: Yes, indeed. Thank you for your detailed review.

**Comment by hirochachacha** [comment]
rebased.
I did `git pull --rebase origin pull/3255/head` and `git pull --rebase origin master`.
Because test is broken currently. https://github.com/aws-amplify/amplify-cli/pull/3255

**Comment by edwardfoyle** [comment]
Hi @hirochachacha thanks for the PR! The only major concern I have with your approach here is that we don't want to introduce a new category for layers. Categories should be broad areas of functionality, not implementation-specific functions. I am working on [another PR](https://github.com/aws-amplify/amplify-cli/pull/3517) that allows the function plugin to get additional features from other util plugins. I think we may want to go that route for this feature. I'm still actively working on that PR so I wouldn't build too many assumptions off of it at this point. But the way it would work is you could make a utility plugin that supplies layers to the function plugin and the function plugin could be updated to include this in the construction of the function.

Also, we probably want to hide questions related to layers under an "advanced options" question in the function walkthrough. We are planning to allow some other configuration of options for less common use cases and we want to put all of that behind this question in the flow to avoid bloating the flow for common use cases

**Comment by hirochachacha** [comment]
@edwardfoyle I'm not sure but your PR seems to try to support multi languages - not only `nodejs`, but also `python`, `ruby`, `jave`, `go`, ...etc. I don't see strong relationship with lambda layer. As you know, [lambda layer](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-layerversion.html) have different entity and different lifecycle. We still need to manage that somehow.

**Comment by edwardfoyle** [comment]
@hirochachacha yes, the main intent of my PR is to add multi-runtimes as a feature, but in doing that I'm refactoring the existing functions package to allow other plugins to supply additional functionality (runtime, templates and perhaps others features in the future including layers) to functions. I agree there may be other things to think about in the case of layers that my PR does not address, but we want to go the route of extending the functionality of the existing functions plugin rather than creating a new category for layers.

Layers are definitely something that we want to support so I'd be happy to discuss implementation options with you

**Comment by hirochachacha** [comment]
@edwardfoyle I see. So I guess you want to address both functions and layers in the same package to standardize package build process like `yarn install`. For example, `amplify-util-node-runtime-provider` handles `nodejs` functions and `nodejs` layers. That's fine. 

I can imagine 3 changes between this PR and your approach.

1. I still don't think treating layers as functions is good idea, they need own lifecycle management. Maybe nested subcommands are preferable. e.g. `amplify function layer add`, `amplify function layer remove`, ... etc. 

2. `backend-config.json` should be extended to distinguish runtimes and layers.
```
	"function": {
		"function1": {
			"service": "Lambda",
			"providerPlugin": "awscloudformation",
			"build": true,
			"runtime": "nodejs",
		},
		"layer1": {
			"service": "LambdaLayer",
			"providerPlugin": "awscloudformation",
			"build": true,
			"runtime": "nodejs",
		},
```

3. Optionally, people might want to split function directory per runtime/layer.
```
amplify/backend/function
			├── java
			├── java_layer
			├── nodejs
			└── nodejs_layer
```

Let me know if you have a different idea.

**Comment by jhockett** [comment]
Hi @hirochachacha. As @edwardfoyle said, thanks for the PR! After some internal debate, we decided against Lambda Layers as a separate category. We're working on an implementation that extends the existing functionality of the Function category. In general, we try to avoid adding new categories unless it's absolutely necessary. I'm going to close this PR, but please feel free to comment below to continue the discussion.

**Comment by hirochachacha** [comment]
@jhockett If you say so, can you please take a look #4112 ?. The only requirement of the PR is using "LambdaLayer" instead of "Lambda" for service name for Lambda Layer support. 
I hope I don't have to wait the next couple of months just for this.
/CC @edwardfoyle

**Pull request closed without merging**